File: Programmers/2019-kakao-recruit/Lv2_오픈채팅방.py
def solution(record):
    message = {'Enter': '님이 들어왔습니다.', 'Leave': '님이 나갔습니다.'}
    info = {}
    answer = []
    
    for m in record:
        m_list = m.split(" ")
        word = m_list[0]
        uid = m_list[1]
        
        if word == 'Enter' or word == 'Change':
            name = m_list[2]
            info[uid] = name
        if word == 'Enter' or word == 'Leave':
            answer.append([uid, message[word]])
    
    return [''.join([info[m[0]], m[1]]) for m in answer]

File: Programmers/2019-kakao-recruit/Lv2_오픈채팅방.js
function solution(record) {
  const answer = [];
  const users = {};
  const commands = {
    Enter: (name) => `${name}님이 들어왔습니다.`,
    Leave: (name) => `${name}님이 나갔습니다.`,
  };

  record.forEach((data) => {
    const [cmd, uid, nickname] = data.split(" ");
    if (cmd === "Enter" || cmd === "Change") {
      users[uid] = nickname;
    }
    if (cmd === "Enter" || cmd === "Leave") {
      answer.push([uid, cmd]);
    }
  });

  return answer.map(([uid, cmd]) => commands[cmd](users[uid]));
}
